imports/wasi_snapshot_preview1/poll.go

```diff
@@ -72,88 +80,161 @@ func pollOneoffFn(ctx context.Context, mod api.Module, params []uint64) syscall.
 
 	// Loop through all subscriptions and write their output.
 
+	// Extract FS context, used in the body of the for loop for FS access.
+	fsc := mod.(*wasm.ModuleInstance).Sys.FS()
+	// Slice of events that are processed out of the loop (stdin subscribers).
+	var stdinSubs []*event
+	// The timeout is initialized at max Duration, the loop will find the minimum.
+	var timeout time.Duration = 1<<63 - 1
+	// Count of all the subscribers that have been already written back to outBuf.
+	readySubs := 0
+
 	// Layout is subscription_u: Union
 	// https://github.com/WebAssembly/WASI/blob/snapshot-01/phases/snapshot/docs.md#subscription_u
 	for i := uint32(0); i < nsubscriptions; i++ {
 		inOffset := i * 48
 		outOffset := i * 32
 
 		eventType := inBuf[inOffset+8] // +8 past userdata
-		var errno syscall.Errno        // errno for this specific event (1-byte)
+		// +8 past userdata +8 contents_offset
+		argBuf := inBuf[inOffset+8+8:]
+		userData := inBuf[inOffset : inOffset+8]
+
+		evt := &event{
+			eventType: eventType,
+			userData:  userData,
+			errno:     wasip1.ErrnoSuccess,
+			outOffset: outOffset,
+		}
+
 		switch eventType {
 		case wasip1.EventTypeClock: // handle later
-			// +8 past userdata +8 contents_offset
-			errno = processClockEvent(ctx, mod, inBuf[inOffset+8+8:])
-		case wasip1.EventTypeFdRead, wasip1.EventTypeFdWrite:
-			// +8 past userdata +8 contents_offset
-			errno = processFDEvent(mod, eventType, inBuf[inOffset+8+8:])
+			newTimeout, err := processClockEvent(argBuf)
+			if err != 0 {
+				return err
+			}
+			// Min timeout.
+			if newTimeout < timeout {
+				timeout = newTimeout
+			}
+			// Ack the clock event to the outBuf.
+			writeEvent(outBuf, evt)
+		case wasip1.EventTypeFdRead:
+			fd := le.Uint32(argBuf)
+			if fd == internalsys.FdStdin {
+				// if the fd is Stdin, do not ack yet,
+				// append to a slice for delayed evaluation.
+				stdinSubs = append(stdinSubs, evt)
+			} else {
+				evt.errno = processFDEventRead(fsc, fd)
+				writeEvent(outBuf, evt)
```

Comment:
Related to https://github.com/tetratelabs/wazero/pull/1346#discussion_r1162492497

When calling `poll_oneoff` multiple times it seems to return the result of the previous call.

sample c code:

```c
#include <stdio.h>
#include <unistd.h>
#include <sys/select.h>

int main(int argc, char **argv)
{
  char buf[5];
  int len = 5, ret;
  fd_set rfds;
  struct timeval tv;

  for (int i = 0; i < 10; i++) {
    printf("[%d] checking inputs ------------ \n", i);
    fflush(stdout);
    FD_ZERO(&rfds);
    FD_SET(0, &rfds);
    tv.tv_sec = 1;
    tv.tv_usec = 0;
    ret = select(1, &rfds, NULL, NULL, &tv);
    if ((ret > 0) && FD_ISSET(0, &rfds)) {
      printf("input detected %d\n", ret);
      ret = read(0, buf, len);
      if (ret > 0) {
        printf("got %.*s (first 5 chars)\n", ret, buf);
      }
    } else {
      printf("no input: %d\n", ret);
    }
  }
}
```

For debugging, we can inspect the result contents of `poll_oneoff` by the following patch:

```diff
diff --git a/imports/wasi_snapshot_preview1/poll.go b/imports/wasi_snapshot_preview1/poll.go
index 80ed4873..4e29fa24 100644
--- a/imports/wasi_snapshot_preview1/poll.go
+++ b/imports/wasi_snapshot_preview1/poll.go
@@ -4,6 +4,7 @@ import (
        "context"
        "syscall"
        "time"
+       "fmt"
 
        "github.com/tetratelabs/wazero/api"
        "github.com/tetratelabs/wazero/internal/platform"
@@ -71,6 +72,10 @@ func pollOneoffFn(ctx context.Context, mod api.Module, params []uint64) syscall.
        if !ok {
                return syscall.EFAULT
        }
+       fmt.Println("outBuf value (initial): ", outBuf)
+       defer func() {
+               fmt.Println("outBuf value (result): ", outBuf)
+       }()
 
        // Eagerly write the number of events which will equal subscriptions unless
        // there's a fault in parsing (not processing).
```

As shown in the following, when an input happens on a call of `select`, next invocation of `select` also falsely detects the input. `poll_oneoff` seems to returns the the previous call's result which is contained in the uninitialized buffer passed as the argument.

```console
$  go run main.go ./out/out.wasm
[0] checking inputs ------------ 
outBuf value (initial):  [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 202 154 59 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
outBuf value (result):  [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 202 154 59 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
no input: 0
[1] checking inputs ------------ 
outBuf value (initial):  [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 202 154 59 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
a
outBuf value (result):  [0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 202 154 59 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
input detected 1
got a
 (first 5 chars)
[2] checking inputs ------------ 
outBuf value (initial):  [0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 202 154 59 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
outBuf value (result):  [0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 202 154 59 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
input detected 1
```

This can be fixed by explicitly initializing the buffer passed by the argument.

```diff
diff --git a/imports/wasi_snapshot_preview1/poll.go b/imports/wasi_snapshot_preview1/poll.go
index 80ed4873..826b6888 100644
--- a/imports/wasi_snapshot_preview1/poll.go
+++ b/imports/wasi_snapshot_preview1/poll.go
@@ -71,6 +71,9 @@ func pollOneoffFn(ctx context.Context, mod api.Module, params []uint64) syscall.
        if !ok {
                return syscall.EFAULT
        }
+       for i := range outBuf {
+               outBuf[i] = 0
+       }
 
        // Eagerly write the number of events which will equal subscriptions unless
        // there's a fault in parsing (not processing).
```